core/transaction/utils.go

```diff
@@ -0,0 +1,400 @@
+			return err
+		}
+
+		err = verifyMerklePath(err, merklePathRawJSON, txn.Hash, b.MerkleTreeRoot)
+		if err != nil {
+			return err
+		}
+
+		url = req.URL.String()
+		chain = append(chain, b)
+		return nil
+	})
+
+	retries := 0
+	ticker := time.NewTicker(time.Second)
+L:
+	//loop query confirmation
+	for retries < 30 {
+		select {
+		case <-ticker.C:
+			retries++
+			r.DoGet(context.TODO(), urls...)
+			//need single valid confirmation
+			errs := r.First()
+			if len(errs) == 0 {
+				break L
+			}
+		}
+	}
+
+	if len(chain) == 0 {
+		return nil, errors.New("verify", "can't get confirmation after 10 retries")
+	}
+
+	//remove current sharder from the list to avoid building chain with it
+	toDelete := mappedSharders[url]
+	for i, s := range sharders {
+		if s == toDelete {
+			sharders = append(sharders[:i], sharders[i+1:]...)
+			break
+		}
+	}
+
+	//build blockchain starting from confirmation block
+	curRound := chain[0].Round
+	rb := resty.New(options...).Then(func(req *http.Request, resp *http.Response, respBody []byte, cf context.CancelFunc, err error) error {
+		if err != nil { //network issue
+			return err
+		}
+
+		if resp.StatusCode != 200 {
+			return errors.Throw(ErrInvalidRequest, strconv.Itoa(resp.StatusCode)+": "+resp.Status)
+		}
+
+		curBlock := &Block{}
+		err = json.Unmarshal(respBody, &curBlock)
+		if err != nil {
+			return err
+		}
+
+		//get tail block and check that current extends it
+		prevBlock := chain[len(chain)-1]
+		if prevBlock.Hash == curBlock.PrevHash && prevBlock.Round+1 == curBlock.Round {
+			blockHeader := &RoundBlockHeader{
+				Version:               curBlock.Version,
+				CreationDate:          curBlock.CreationDate,
+				Hash:                  curBlock.Hash,
+				PreviousBlockHash:     curBlock.PrevHash,
+				MinerID:               curBlock.MinerID,
+				Round:                 curBlock.Round,
+				RoundRandomSeed:       curBlock.RoundRandomSeed,
+				MerkleTreeRoot:        curBlock.MerkleTreeRoot,
+				StateChangesCount:     0,
+				StateHash:             curBlock.StateHash,
+				ReceiptMerkleTreeRoot: curBlock.ReceiptMerkleTreeRoot,
+				NumberOfTxns:          int64(curBlock.NumTxns),
+			}
+			err = validateBlockHash(blockHeader)
+			if err != nil {
+				return err
+			}
+
+			chain = append(chain, blockHeader)
+			return nil
+		}
+		return errors.New("get_block", "wrong block")
+	})
+
+	//query for blocks until ConfirmationChainLength is built or every sharder is queried
+	for len(chain) < cfg.ConfirmationChainLength && len(sharders) > 0 {
+		//for every new block create sharder list to query
+		rand := util.NewRand(len(sharders))
+		//iterate through all sharders sequentially to get next block
+		for {
+			next, err := rand.Next()
+			if err != nil {
+				return nil, errors.New("get_round_block", "can't get round block, blockchain might be stuck")
+			}
+
+			cur := sharders[next]
+			burl := fmt.Sprintf("%v/%v%v", cur, BLOCK_BY_ROUND_URL, curRound+1)
+			rb.DoGet(context.TODO(), burl)
```

Comment:
Not sure if the resty instance could be reused without concurrent issue. We'd better check

Reply:
this is done also in previous implementation ``func VerifyTransaction(txnHash string, sharders []string) (*Transaction, error) {
``, so if we need to change it we need to do it in both places, can we do it in different issue?

Reply:
Sure, let's do in another issue if needed.